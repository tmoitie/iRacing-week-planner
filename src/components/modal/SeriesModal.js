// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import classnames from 'classnames';
import intersection from 'lodash.intersection';
import Modal from './Modal';
import RaceLength from '../columns/RaceLength';

import allRaces from '../../lib/races';
import offWeekData from '../../data/offWeeks';

import styles from '../../styles/main.module.scss';

type Props = {
  onClose: () => void,
  isOpen: boolean,
  ownedTracks: Array<number>,
  ownedCars: Array<number>,
  seriesId: number,
  seasonId: number,
};

export default function SeriesModal({ onClose, ownedTracks, ownedCars, isOpen, seriesId, seasonId }: Props) {
  const races = allRaces.filter((race) => race.seriesId === seriesId);
  const { t } = useTranslation();

  const now = moment().utc();

  const dateParams = {
    date: {
      dateStyle: 'long',
      timeZone: 'UTC',
    },
  };

  const showCar = !!(offWeekData[seasonId]?.carByWeek);

  return (
    <Modal
      id={`modal-series-${seriesId}-${seasonId}`}
      isOpen={isOpen}
      onClose={onClose}
      title={t('Tracks for {{series}}', {
        series: t(races[0].series),
      })}
      doneAction={onClose}
    >
      <div className={styles['container-fluid']}>
        <div className={styles['table-responsive']}>
          <table className={styles.table} style={{ fontSize: '0.8em' }}>
            <thead>
              <tr>
                <th>{t('Week')}</th>
                <th>{t('Track')}</th>
                {showCar ? <th>{t('Car')}</th> : null}
                <th>{t('Start')}</th>
                <th>{t('End')}</th>
                <th>{t('Length')}</th>
              </tr>
            </thead>
            <tbody>
              {races.map((race) => {
                const raceWeekEnd = moment(race.startTime).add(race.weekLength);
                const current = now.isBetween(race.startTime, raceWeekEnd);

                const startDate = moment(race.startTime).toDate();
                const endDate = moment(race.startTime).add(race.weekLength)
                  .subtract(1, 'days')
                  .toDate();

                const knownCar = race.carIds[0] !== null;

                return (
                  <tr key={race.week} style={current ? { fontWeight: 700 } : {}}>
                    <td>
                      {race.week + 1}
                    </td>
                    <td className={classnames({ [styles.success]: ownedTracks.indexOf(race.trackId) !== -1 })}>
                      {t(race.track)}
                    </td>
                    {showCar ? (
                      <td
                        className={classnames({
                          [styles.success]: knownCar ? intersection(ownedCars, race.carIds).length !== 0 : null,
                        })}
                      >
                        {knownCar ? race.carClasses.join(', ') : t('Unknown')}
                      </td>
                    ) : null}
                    <td>
                      {t('{{date, datetime}}', { date: startDate, formatParams: dateParams })}
                    </td>
                    <td>
                      {t('{{date, datetime}}', { date: endDate, formatParams: dateParams })}
                    </td>
                    <RaceLength race={race} />
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    </Modal>
  );
}
