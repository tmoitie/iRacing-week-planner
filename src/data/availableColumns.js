// @flow

import type { ElementType } from 'react';
import moment from 'moment';
import {
  Id, Car, Class, EndDate, Fixed, Licence, LinkColumn, NextRace, Official,
  RaceTimes, Series, SeasonEnd, StartDate, Track, Type,
} from '../components/columns';
import RaceLength from '../components/columns/RaceLength';
import { getNextRace } from '../lib/races';
import type { SeriesRace } from '../lib/races';

/* eslint react/no-multi-comp: 0 */

const defaultSort = (order: 'asc' | 'desc', a: SeriesRace, b: SeriesRace) => {
  if (a.licenceClassNumber === b.licenceClassNumber) {
    if (a.series === b.series) {
      return 0;
    }

    if (order === 'asc') {
      return a.series.toLowerCase() < b.series.toLowerCase() ? -1 : 1;
    }

    return a.series.toLowerCase() < b.series.toLowerCase() ? 1 : -1;
  }

  if (order === 'asc') {
    return a.licenceClassNumber < b.licenceClassNumber ? -1 : 1;
  }

  return a.licenceClassNumber < b.licenceClassNumber ? 1 : -1;
};

const sortByDate = (order: 'asc' | 'desc', a?: moment$Moment, b?: moment$Moment) => {
  if (a === b) {
    return 0;
  }
  if (a === undefined) {
    return 1;
  }
  if (b === undefined) {
    return -1;
  }
  if (a.isSame(b)) {
    return 0;
  }
  if (order === 'asc') {
    return a.isBefore(b) ? -1 : 1;
  }

  return a.isAfter(b) ? -1 : 1;
};

const getSortByDate = (key: string) => (order: 'asc' | 'desc', a: SeriesRace, b: SeriesRace) => {
  const sort = sortByDate(order, a[key], b[key]);
  return sort !== 0 ? sort : defaultSort(order, a, b);
};

export type ColumnType = {
  id: string,
  header: string,
  component: ElementType,
  default?: boolean,
  forced?: boolean,
  sort?: (order: 'asc' | 'desc', a: SeriesRace, b: SeriesRace) => number,
};

const availableColumns: Array<ColumnType> = [{
  id: 'id',
  header: 'ID',
  component: Id,
  default: false,
  sort: (order, a, b) => {
    if (a.seriesId === b.seriesId) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (a.seriesId < b.seriesId ? -1 : 1);
    }
    return (a.seriesId > b.seriesId ? -1 : 1);
  },
}, {
  id: 'class',
  header: 'Class',
  component: Class,
  default: true,
  sort: (order, a, b) => defaultSort(order, a, b),
}, {
  id: 'licence',
  header: 'Licence',
  component: Licence,
  default: true,
  sort: (order, a, b) => {
    if (a.licenceLevel === b.licenceLevel) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return a.licenceLevel < b.licenceLevel ? -1 : 1;
    }
    return a.licenceLevel > b.licenceLevel ? -1 : 1;
  },
}, {
  id: 'type',
  header: 'Type',
  component: Type,
  default: true,
  sort: (order, a, b) => {
    if (a.type === b.type) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (a.type < b.type ? -1 : 1);
    }
    return (a.type > b.type ? -1 : 1);
  },
}, {
  id: 'series',
  header: 'Series',
  component: Series,
  default: true,
  forced: true,
  sort: (order, a, b) => {
    if (a.series === b.series) {
      return 0;
    }
    if (order === 'asc') {
      return (a.series.toLowerCase() < b.series.toLowerCase() ? -1 : 1);
    }
    return (a.series.toLowerCase() > b.series.toLowerCase() ? -1 : 1);
  },
}, {
  id: 'track',
  header: 'Track',
  component: Track,
  default: true,
  sort: (order, a, b) => {
    if (a.track === b.track) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (a.track.toLowerCase() < b.track.toLowerCase() ? -1 : 1);
    }
    return (a.track.toLowerCase() > b.track.toLowerCase() ? -1 : 1);
  },
}, {
  id: 'car',
  header: 'Car',
  component: Car,
  default: true,
  sort: (order, a, b) => {
    const carA = a.carClasses.join('');
    const carB = b.carClasses.join('');
    if (carA === carB) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (carA.toLowerCase() < carB.toLowerCase() ? -1 : 1);
    }
    return (carA.toLowerCase() > carB.toLowerCase() ? -1 : 1);
  },
}, {
  id: 'start',
  header: 'Start',
  component: StartDate,
  default: true,
  sort: getSortByDate('startTime'),
}, {
  id: 'end',
  header: 'End',
  component: EndDate,
  sort: (order, a, b) => {
    const timeA = moment(a.startTime).add(a.weekLength);
    const timeB = moment(b.startTime).add(b.weekLength);
    if (timeA.isSame(timeB)) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return timeA.isBefore(timeB) ? -1 : 1;
    }

    return timeA.isAfter(timeB) ? -1 : 1;
  },
}, {
  id: 'racelength',
  header: 'Length',
  component: RaceLength,
  default: true,
}, {
  id: 'official',
  header: 'Official',
  component: Official,
  default: true,
  sort: (order, a, b) => {
    if (a.official === b.official) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (a.official === false ? -1 : 1);
    }
    return (a.official === true ? -1 : 1);
  },
}, {
  id: 'fixed',
  header: 'Fixed',
  component: Fixed,
  sort: (order, a, b) => {
    if (a.fixed === b.fixed) {
      return defaultSort(order, a, b);
    }
    if (order === 'asc') {
      return (a.fixed === false ? -1 : 1);
    }
    return (a.fixed === true ? -1 : 1);
  },
}, {
  id: 'raceTimes',
  header: 'Race times',
  component: RaceTimes,
}, {
  id: 'nextRace',
  header: 'Next race',
  component: NextRace,
  default: true,
  sort: (order, a, b) => {
    const now = moment().utc();
    const aNextDate = getNextRace(now, a);
    const bNextDate = getNextRace(now, b);
    if (aNextDate === null && bNextDate === null) {
      return defaultSort(order, a, b);
    }
    if (aNextDate === null) {
      return order === 'asc' ? 1 : -1;
    }
    if (bNextDate === null) {
      return order === 'asc' ? -1 : 1;
    }
    const sort = sortByDate(order, aNextDate, bNextDate);

    return sort !== 0 ? sort : defaultSort(order, a, b);
  },
}, {
  id: 'seriesEnd',
  header: 'Season end',
  component: SeasonEnd,
  default: false,
  sort: getSortByDate('seriesEnd'),
}, {
  id: 'seriesLink',
  header: 'Link',
  component: LinkColumn,
  default: true,
}];

export default availableColumns;
