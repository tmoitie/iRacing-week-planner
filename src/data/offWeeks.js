export default {
  // Commented out series kept in here for when iRacing schedule next season, often with the same settings
  // 512 │ 3974 | SCCA Runoffs - FR500
  // 3974: {
  //   weekEndOffset: { days: 7 },
  // },
  // 506 │ 3968 | SCCA Runoffs - Formula Vee
  // 3968: {
  //   weekEndOffset: { days: 7 },
  // },
  // 509 │ 3971 | SCCA Runoffs - GT1
  // 3971: {
  //   weekEndOffset: { days: 7 },
  // },
  // 511 │ 3973 | SCCA Runoffs - GT4
  // 3973: {
  //   weekEndOffset: { days: 7 },
  // },
  // 508 │ 3970 | SCCA Runoffs - Global MX-5 Spec Miata
  // 3970: {
  //   weekEndOffset: { days: 7 },
  // },
  // 507 │ 3969 | SCCA Runoffs - Spec Racer Ford
  // 3969: {
  //   weekEndOffset: { days: 7 },
  // },
  // 513 │ 3975 | SCCA Runoffs - Touring Cars
  // 3975: {
  //   weekEndOffset: { days: 7 },
  // },
  // 510 │ 3972 | SCCA Runoffs - USF2000
  // 3972: {
  //   weekEndOffset: { days: 7 },
  // },
  // 478 │ 3965 │ eNASCAR RTP Contender Series
  // 3965: {
  //   weekStartOffset: { days: 6 },
  //   offWeeks: [
  //     3.1,
  //   ],
  // },
  // 229 │ 4113 │ NASCAR iRacing Series - Open | 2023-11-07
  // 4113: {
  //   offWeeks: [
  //     13.1, 16.1,
  //   ],
  // },
  // 437 │ 4119 │ BMW M Sim Cup | 2023-11-21
  // 4119: {
  //   offWeeks: [
  //     1.1, 1.1, 1.1, 1.1, 1.1, 2.1, 2.1, 2.1, 2.1, 3.1, 4.1, 4.1, 4.1, 5.1, 5.1, 5.1, 6.1, 6.1, 6.1, 6.1, 6.1, 6.1, 7.1,
  //     7.1, 7.1, 7.1, 7.1, 7.1, 7.1, 8.1, 8.1, 9.1,
  //   ],
  // },
  // 469 │ 4315 | Global Endurance Pure Driving School Series | 2023-09-05
  // 4315: {
  //   weekStartOffset: { days: 7 },
  //   offWeeks: [
  //     1.1, 2.1, 3.1, 4.1, 5.1,
  //   ],
  // },
  // 439 │ 4384 │ eNASCAR Road to Pro Qual Series - Round 2 | 2023-10-03
  // 4384: {
  //   offWeeks: [
  //     1.1, 2.1, 3.1, 4.1, 5.1, 6.1,
  //   ],
  // },
  // 307 │ 4485 │ Dirt │ World of Outlaws Sprint Car Series | 2023-12-05
  // 4485: {
  //  weekStartOffset: { days: -1 },
  // },
  // 478 │ 4538   │ Oval │ eNASCAR RTP Contender Series
  // 4538: {
  //   weekStartOffset: { days: 6 },
  //   offWeeks: [
  //     6.1,
  //   ],
  // },
  // 526 │ 4666   │ Road │ C     │ Ring Meister Ricmotech Series - Fixed | 2024-03-05
  // 4666: {
  //   carByWeek: [
  //     [10009], // 1 Skip Barber Formula 2000
  //     [10567], // 2 Audi R8 LMS EVO II GT3
  //     [10516], // 3 Hyundai Elantra N TC
  //     [10232], // 4 Street Stock
  //     [10484], // 5 Chevrolet Corvette C8.R GTE
  //     [10497], // 6 McLaren 570S GT4
  //     [10011], // 7 Pontiac Solstice
  //     [10040], // 8 Radical SR8
  //     [10552], // 9 Ligier JS P320
  //     [10006], // 10 [Legacy] Pro Mazda
  //     [10473], // 11 Porsche 718 Cayman GT4 Clubsport MR
  //     [10495], // 12 Lamborghini Huracán GT3 EVO
  //   ],
  //   carClassShortNameByWeek: [
  //     ['SBRS'], // 1 Skip Barber Formula 2000
  //     ['Audi R8 EVO2 GT3'], // 2 Audi R8 LMS EVO II GT3
  //     ['Elantra N TC'], // 3 Hyundai Elantra N TC
  //     ['Street Stock'], // 4 Street Stock
  //     ['Vette C8.R GTE'], // 5 Chevrolet Corvette C8.R GTE
  //     ['McLaren 570S GT4'], // 6 McLaren 570S GT4
  //     ['Solstice'], // 7 Pontiac Solstice
  //     ['SR8'], // 8 Radical SR8
  //     ['Ligier JS P320'], // 9 Ligier JS P320
  //     ['Pro Mazda'], // 10 [Legacy] Pro Mazda
  //     ['Porsche 718'], // 11 Porsche 718 Cayman GT4 Clubsport MR
  //     ['Lambo Huracán'], // 12 Lamborghini Huracán GT3 EVO
  //   ],
  // },
  // 538 │ 4667   │ Oval │ D     │ Draft Master - Fixed | 2024-03-05
  // 4667: {
  //   carByWeek: [
  //     [10481, 10530, 10459], // 1 NASCAR Trucks
  //     [10068], // 2 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
  //     [10464, 10465, 10466], // 3 NASCAR Xfinity Cars
  //     [10527, 10482, 10483, 10566], // 4 NASCAR Classic 1987
  //     [10088], // 5 Gen 4 Cup
  //     [10503, 10504, 10505], // 6 NASCAR Next Gen
  //     [10481, 10530, 10459], // 7 NASCAR Trucks
  //     [10068], // 8 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
  //     [10464, 10465, 10466], // 9 NASCAR Xfinity Cars
  //     [10527, 10482, 10483, 10566], // 10 NASCAR Classic 1987
  //     [10088], // 11 Gen 4 Cup
  //     [10503, 10504, 10505], // 12 NASCAR Next Gen
  //   ],
  //   carClassShortNameByWeek: [
  //     ['NASCAR Trucks'], // 1 NASCAR Trucks
  //     ['[Legacy] NASCAR Cup Chevrolet Impala COT - 2009'], // 2 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
  //     ['XFINITY Series'], // 3 NASCAR Xfinity Cars
  //     ['NASCAR 1987'], // 4 NASCAR Classic 1987
  //     ['Gen 4 Cup'], // 5 Gen 4 Cup
  //     ['NASCAR 2022 NG'], // 6 NASCAR Next Gen
  //     ['NASCAR Trucks'], // 7 NASCAR Trucks
  //     ['[Legacy] NASCAR Cup Chevrolet Impala COT - 2009'], // 8 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
  //     ['XFINITY Series'], // 9 NASCAR Xfinity Cars
  //     ['NASCAR 1987'], // 10 NASCAR Classic 1987
  //     ['Gen 4 Cup'], // 11 Gen 4 Cup
  //     ['NASCAR 2022 NG'], // 12 NASCAR Next Gen
  //   ],
  // },
  // 207 │ 4703 │ NASCAR iRacing Series - Fixed | 2024-11-12
  4703: {
    offWeeks: [
      13.1, 22.1, 22.1,
    ],
  },
  // 229 │ 4699 │ NASCAR iRacing Series - Open | 2024-11-12
  4699: {
    offWeeks: [
      13.1, 22.1, 22.1,
    ],
  },
  // 438 │ 4704   │ Oval │ A     │ eNASCAR Road to Pro Qual Series - Round 1 | 2024-05-28
  4704: {
    offWeeks: [
      1.1, 2.1, 3.1, 4.1, 5.1, 6.1, 7.1,
    ],
  },
  // 495 │ 4828   │ Road │ C     │ Formula A - Grand Prix Tour | 2024-12-10
  4828: {
    offWeeks: [
      2.1, 3.1, 4.1, 5.1, 6.1, 8.1, 9.1, 12.1, 14.1, 14.1, 14.1, 16.1, 18.1, 18.1, 18.1, 21.1, 21.1,
    ],
  },
  // 496 │ 4829   │ Road │ C     │ Formula A - Grand Prix Tour - Fixed │ 2024-12-10
  4829: {
    offWeeks: [
      2.1, 3.1, 4.1, 5.1, 6.1, 8.1, 9.1, 12.1, 14.1, 14.1, 14.1, 16.1, 18.1, 18.1, 18.1, 21.1, 21.1,
    ],
  },
  // 374 │ 4831   │ Road │ C     │ INDYCAR iRacing Series | 2024-09-17
  4831: {
    offWeeks: [
      1.1, 2.1, 2.1, 2.1, 4.1, 5.1, 8.1, 9.1, 12.1, 12.1, 12.1, 15.1,
    ],
  },
  // 538 │ 4822   │ Oval        │ D     │ Draft Master - Fixed | 2024-06-04
  4822: {
    carByWeek: [
      [10068], // 1 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      [10481, 10530, 10459], // 2 NASCAR Trucks
      [10527, 10482, 10483, 10566], // 3 NASCAR Classic 1987
      [10503, 10504, 10505], // 4 NASCAR Next Gen
      [10068], // 5 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      [10481, 10530, 10459], // 6 NASCAR Trucks
      [10527, 10482, 10483, 10566], // 7 NASCAR Classic 1987
      [10503, 10504, 10505], // 8 NASCAR Next Gen
      [10068], // 9 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      [10481, 10530, 10459], // 10 NASCAR Trucks
      [10527, 10482, 10483, 10566], // 11 NASCAR Classic 1987
      [10503, 10504, 10505], // 12 NASCAR Next Gen
    ],
    carClassShortNameByWeek: [
      ['[Legacy] NASCAR Cup Chevrolet Impala COT - 2009'], // 1 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      ['NASCAR Trucks'], // 2 NASCAR Trucks
      ['NASCAR 1987'], // 3 NASCAR Classic 1987
      ['NASCAR 2022 NG'], // 4 NASCAR Next Gen
      ['[Legacy] NASCAR Cup Chevrolet Impala COT - 2009'], // 5 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      ['NASCAR Trucks'], // 6 NASCAR Trucks
      ['NASCAR 1987'], // 7 NASCAR Classic 1987
      ['NASCAR 2022 NG'], // 8 NASCAR Next Gen
      ['[Legacy] NASCAR Cup Chevrolet Impala COT - 2009'], // 9 [Legacy] NASCAR Cup Chevrolet Impala COT - 2009
      ['NASCAR Trucks'], // 10 NASCAR Trucks
      ['NASCAR 1987'], // 11 NASCAR Classic 1987
      ['NASCAR 2022 NG'], // 12 NASCAR Next Gen
    ],
  },
  // 526 │ 4820   │ Sports Car  │ C     │ Ring Meister Ricmotech Series - Fixed | 2024-06-04
  4820: {
    carByWeek: [
      [10306], // 1 Cadillac CTS-V Racecar
      [10548], // 2 Renault Clio
      [10552], // 3 Ligier JS P320
      [10110], // 4 Lotus 79
      [10517], // 5 Honda Civic Type R TCR
      [10457], // 6 BMW M8 GTE
      [10524], // 7 Stock Car Brasil Chevrolet Cruze
      [10394], // 8 Global Mazda MX-5 Cup
      [10113], // 9 Chevrolet Corvette C6.R GT
      [10563], // 10 Ferrari 296 GT3
      [10040], // 11 Radical SR8
      [10507], // 12 Formula Vee
    ],
    carClassShortNameByWeek: [
      ['Cadillac CTS-VR'], // 1 Cadillac CTS-V Racecar
      ['Renault Clio'], // 2 Renault Clio
      ['Ligier JS P320'], // 3 Ligier JS P320
      ['Lotus 79'], // 4 Lotus 79
      ['Civic Type R'], // 5 Honda Civic Type R TCR
      ['BMW M8 GTE'], // 6 BMW M8 GTE
      ['Stock Car Brasil'], // 7 Stock Car Brasil Chevrolet Cruze
      ['MX5 Cup 2016'], // 8 Global Mazda MX-5 Cup
      ['C6R GT1'], // 9 Chevrolet Corvette C6.R GT
      ['Ferrari 296 GT3'], // 10 Ferrari 296 GT3
      ['SR8'], // 11 Radical SR8
      ['Formula Vee'], // 12 Formula Vee
    ],
  },
  // 557 │ 4825   │ Sports Car  │ D     │ BMW M Power Tour - Fixed | 2024-06-04
  4825: {
    weekEndOffset: { days: 7 },
  },
  // 331 │ 4778   │ Sports Car  │ B     │ Global Endurance CONSPIT Tour  | 2024-06-04
  4778: {
    weekStartOffset: { days: 7 },
    offWeeks: [
      1.1, 2.1, 3.1, 4.1, 5.1,
    ],
  },
  // 419 │ 4782   │ Sports Car  │ B     │ IMSA Endurance Series  | 2024-05-28
  4782: {
    offWeeks: [
      1.1, 2.1, 3.1, 4.1, 5.1,
    ],
  },
  // 501 │ 4806   │ Dirt        │ C     │ Dirt 410 Sprint Car Tour - Fixed | 2024-06-04 | recheck - missing weeks
  4806: {
    offWeeks: [
      0.1, 0.1,
    ],
  },
  // 275 │ 4862   │ Sports Car  │ D     │ Nurburgring Endurance Championship | 2024-10-29
  4862: {
    offWeeks: [
      1.1, 1.1, 2.1, 2.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 4.1, 5.1, 5.1, 6.1, 7.1, 7.1, 7.1, 8.1, 8.1, 8.1,
    ],
  },

  // 451 │ 4861   │ Sports Car  │ C     │ Creventic Endurance Series | 2024-12-17
  4861: {
    offWeeks: [
      1.1, 1.1, 1.1, 1.1, 1.1, 1.1, 1.1, 1.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1,
      2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 2.1, 3.1, 3.1, 3.1, 3.1,
    ],
  },

  // 492 │ 4755   │ Sports Car  │ C     │ IMSA Michelin Pilot Challenge Series | 2024-06-04 | recheck - missing weeks
  4755: {
    offWeeks: [
      0.1, 0.1, 1.1, 2.1, 3.1, 4.1, 5.1,
    ],
  },
};
