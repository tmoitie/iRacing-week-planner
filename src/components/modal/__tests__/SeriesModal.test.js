import { describe, test } from '@jest/globals';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MockDate from 'mockdate';
import SeriesModal from '../SeriesModal';
import '../../../data/offWeeks';

jest.mock('firebase/auth');
jest.mock('../../../data/offWeeks');

describe('components/modal/SeriesModal', () => {
  test('renders closed', async () => {
    const onClose = jest.fn(() => {});
    const { baseElement } = render(
      <SeriesModal isOpen={false} onClose={onClose} ownedTracks={[]} ownedCars={[]} seasonId={2500} seriesId={245} />,
    );

    expect(screen.queryByTestId('modal-series-245-2500')).not.toBeTruthy();
  });

  test('renders open', async () => {
    const user = userEvent.setup();

    MockDate.set('2022-09-07T13:30:00.000Z');
    const onClose = jest.fn(() => {});

    const { baseElement } = render(
      <SeriesModal isOpen onClose={onClose} ownedTracks={[6]} ownedCars={[]} seasonId={2500} seriesId={245} />,
    );

    expect(await screen.findByTestId('modal-series-245-2500')).toMatchSnapshot();

    await user.click(screen.getByText('Close'));
    expect(onClose).toHaveBeenCalled();
  });

  test('renders with cars', async () => {
    MockDate.set('2022-09-07T13:30:00.000Z');
    const onClose = jest.fn(() => {});
    const { baseElement } = render(
      <SeriesModal isOpen onClose={onClose} ownedTracks={[6]} ownedCars={[]} seasonId={2550} seriesId={245} />,
    );

    expect(await screen.findByTestId('modal-series-245-2550')).toMatchSnapshot();
  });
});
