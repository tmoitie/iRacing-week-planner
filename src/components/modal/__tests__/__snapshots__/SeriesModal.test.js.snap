// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/SeriesModal renders open 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="modal-series-245-2500"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Tracks for 13th Week iRacing Figure GR8
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <div
          class="table-responsive"
        >
          <table
            class="table"
            style="font-size: 0.8em;"
          >
            <thead>
              <tr>
                <th>
                  Week
                </th>
                <th>
                  Track
                </th>
                <th>
                  Start
                </th>
                <th>
                  End
                </th>
                <th>
                  Length
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  1
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td>
                  September 6, 2022
                </td>
                <td>
                  September 6, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr
                style="font-weight: 700;"
              >
                <td>
                  2
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td>
                  September 7, 2022
                </td>
                <td>
                  September 7, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  3
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td>
                  September 8, 2022
                </td>
                <td>
                  September 8, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  4
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td>
                  September 9, 2022
                </td>
                <td>
                  September 9, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  5
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td>
                  September 10, 2022
                </td>
                <td>
                  September 10, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  6
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td>
                  September 11, 2022
                </td>
                <td>
                  September 11, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  7
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td>
                  September 12, 2022
                </td>
                <td>
                  September 12, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;

exports[`components/modal/SeriesModal renders with cars 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="modal-series-245-2550"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Tracks for 13th Week iRacing Figure GR8
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <div
          class="table-responsive"
        >
          <table
            class="table"
            style="font-size: 0.8em;"
          >
            <thead>
              <tr>
                <th>
                  Week
                </th>
                <th>
                  Track
                </th>
                <th>
                  Car
                </th>
                <th>
                  Start
                </th>
                <th>
                  End
                </th>
                <th>
                  Length
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  1
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 6, 2022
                </td>
                <td>
                  September 6, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr
                style="font-weight: 700;"
              >
                <td>
                  2
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 7, 2022
                </td>
                <td>
                  September 7, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  3
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 8, 2022
                </td>
                <td>
                  September 8, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  4
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 9, 2022
                </td>
                <td>
                  September 9, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  5
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 10, 2022
                </td>
                <td>
                  September 10, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  6
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight Jump
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 11, 2022
                </td>
                <td>
                  September 11, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  7
                </td>
                <td
                  class="success"
                >
                  Irwindale Speedway - Figure Eight
                </td>
                <td
                  class=""
                >
                  Rallycross, Pro 2 Truck
                </td>
                <td>
                  September 12, 2022
                </td>
                <td>
                  September 12, 2022
                </td>
                <td>
                  <div>
                    25L
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
